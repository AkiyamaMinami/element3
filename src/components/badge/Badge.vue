<script setup lang="ts">
import { computed } from 'vue'
import { isString } from '../../utils/types'

type BadgeType = 'primary' | 'success' | 'warning' | 'danger' | 'info'

interface BadgeProps {
  value?: String | Number
  max?: Number
  isDot?: Boolean
  hidden?: Boolean
  type?: BadgeType
}

const props = withDefaults(defineProps<BadgeProps>(), {
  type: 'danger',
})

function useBadgeContent(props) {
  return computed(() => {
    if (!props.max || isString(props.value))
      return props.value

    if (props.value <= props.max)
      return props.value

    return `${props.max}+`
  })
}

const badgeContent = useBadgeContent(props)

function useShow(props) {
  return computed(() => {
    if (props.hidden)
      return false

    if (props.value === 0)
      return false

    return true
  })
}

const isShow = useShow(props)
</script>

<template>
  <div class="el-badge">
    <slot />
    <sup
      v-if="isShow" class="el-badge__content" :class="[
        `el-badge__content--${type}`,
        {
          'is-dot': isDot,
          'is-fixed': $slots.default,
        },
      ]" v-text="badgeContent"
    />
  </div>
</template>

<style lang="scss">
@import '../../theme/src/badge.scss';
</style>
