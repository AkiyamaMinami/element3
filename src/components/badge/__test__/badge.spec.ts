import { mount } from '@vue/test-utils'
import Badge from '../Badge.vue'

describe('Badge.vue', () => {
  it('snapshot', () => {
    const wrapper = mount(Badge)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render props.value', () => {
    const vm = mount(Badge, {
      props: {
        value: 7,
      },
    })

    expect(vm.find('.el-badge__content').text()).toBe('7')
  })

  it('should render string props.value', () => {
    const vm = mount(Badge, {
      props: {
        value: 'Badge',
      },
    })

    expect(vm.find('.el-badge__content').text()).toBe('Badge')
  })

  it('should render props.value when value larger then max', () => {
    const vm = mount(Badge, {
      props: {
        value: 100,
        max: 10,
      },
    })

    expect(vm.find('.el-badge__content').text()).toBe('10+')
  })

  it('should render props.value when value less then max', () => {
    const vm = mount(Badge, {
      props: {
        value: 9,
        max: 10,
      },
    })

    expect(vm.find('.el-badge__content').text()).toBe('9')
  })

  it('should render dot', () => {
    const vm = mount(Badge, {
      props: {
        value: 7,
        max: 10,
        isDot: true,
      },
    })

    const text = vm.find('.el-badge__content')
    expect(text.classes()).toContain('is-dot')
  })

  it('should not render when set hidden', () => {
    const vm = mount(Badge, {
      props: {
        value: 7,
        max: 10,
        hidden: true,
      },
    })

    expect(vm.find('sup').exists()).toBeFalsy()
  })

  it('should render props.type', () => {
    const vm = mount(Badge, {
      props: {
        value: 7,
        type: 'primary',
      },
    })

    const text = vm.find('.el-badge__content')
    expect(text.classes()).toContain('el-badge__content--primary')
  })
})
