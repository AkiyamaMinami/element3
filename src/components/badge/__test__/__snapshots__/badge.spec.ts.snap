// Vitest Snapshot v1

exports[`Badge.vue > snapshot 1`] = `
<div
  class="el-badge"
>
  
  
  <sup
    class="el-badge__content el-badge__content--danger"
  />
</div>
`;
